src/sm/networkmanager/networkmanager.cpp

```diff
@@ -0,0 +1,756 @@
+    if (err = mNetMonitor->StartInstanceMonitoring(
+            instanceID, network.mNetworkParameters.mIP, network.mDownloadLimit, network.mUploadLimit);
+        !err.IsNone()) {
+        cleanupCNI();
+        cleanupNetworkNamespace();
+        cleanupInstanceCache();
+
+        return err;
+    }
+
+    if (err = CreateHostsFile(networkID, network.mNetworkParameters.mIP, network); !err.IsNone()) {
+        cleanupCNI();
+        cleanupNetworkNamespace();
+        cleanupInstanceCache();
+
+        return err;
+    }
+
+    if (err = CreateResolvConfFile(networkID, network, result.mDNSServers); !err.IsNone()) {
+        cleanupCNI();
+        cleanupNetworkNamespace();
+        cleanupInstanceCache();
+
+        return err;
+    }
+
+    if (err = UpdateInstanceNetworkCache(instanceID, networkID, network.mNetworkParameters.mIP, mHost); !err.IsNone()) {
+        cleanupCNI();
+        cleanupNetworkNamespace();
+        cleanupInstanceCache();
+
+        return err;
+    }
+
+    LOG_DBG() << "Instance added to network: instanceID=" << instanceID << ", networkID=" << networkID;
+
+    return ErrorEnum::eNone;
+}
+
+Error NetworkManager::RemoveInstanceFromNetwork(const String& instanceID, const String& networkID)
+{
+    LOG_DBG() << "Remove instance from network: instanceID=" << instanceID << ", networkID=" << networkID;
+
+    if (auto err = IsInstanceInNetwork(instanceID, networkID); !err.IsNone()) {
+        return err;
+    }
+
+    if (auto err = mNetMonitor->StopInstanceMonitoring(instanceID); !err.IsNone()) {
+        return err;
+    }
+
+    cni::NetworkConfigList netConfig;
+    cni::RuntimeConf       rtConfig;
+
+    netConfig.mName    = networkID;
+    netConfig.mVersion = cni::cVersion;
+
+    rtConfig.mContainerID = instanceID;
+
+    if (auto err = GetNetnsPath(instanceID, rtConfig.mNetNS); !err.IsNone()) {
+        return err;
+    }
+
+    rtConfig.mIfName = cInstanceInterfaceName;
+
+    if (auto err = mCNI->DelNetworkList(netConfig, rtConfig); !err.IsNone()) {
+        return err;
+    }
+
+    if (auto err = DeleteNetworkNamespace(instanceID); !err.IsNone()) {
+        return err;
+    }
+
+    if (auto err = RemoveInstanceFromCache(instanceID, networkID); !err.IsNone()) {
+        return err;
+    }
+
+    LOG_DBG() << "Instance removed from network: instanceID=" << instanceID << ", networkID=" << networkID;
+
+    return ErrorEnum::eNone;
+}
+
+Error NetworkManager::GetNetnsPath(const String& instanceID, String& netnsPath) const
+{
+    LOG_DBG() << "Get network namespace path: instanceID=" << instanceID;
+
+    return GetNetworkNamespacePath(instanceID, netnsPath);
+}
+
+Error NetworkManager::GetInstanceIP(const String& instanceID, const String& networkID, String& ip) const
+{
+    LockGuard lock {mMutex};
+
+    LOG_DBG() << "Get instance IP: instanceID=" << instanceID << ", networkID=" << networkID;
+
+    auto network = mNetworkData.At(networkID);
+    if (!network.mError.IsNone()) {
+        return network.mError;
```

Comment:
General comment for whole file: errors coming from external module should be wrapped.

Reply:
Done